docs/spec.md

```diff
@@ -0,0 +1,269 @@
+
+- `notify: () => void;`
+
+The reason for not choosing this approach is because it rules out true reactive or observable to style programming where the `Host` is effectively an `Observer`. Another approach would be to pass the value in the notification so that `Host` really is an observer of values for that property:
+
+- `notify: (resource: Resource, property: string, value: CalcValue) => void;`
+
+A reason for not choosing this approach would be because it couples invalidation of a resource with its re-computation. If I want to dirty a cell and notify its dependents I must now recompute the cell first to obtain the new value. This may be expensive and the host may not even need the value right now. Additionally, it is possible to support this observer style signature using the currently proposed one with a wrapper function:
+
+```ts
+interface Observer {
+  notify: (resource: Resource, property: string, value: CalcValue) => void;
+}
+
+interface Host {
+  notify: (resource: Resource, property) => void;
+}
+
+function observerToHost(observer: Observer): Host {
+  const host = {
+    notify(resource: Resource, property: string) {
+      resource.request(
+        host,
+        property,
+        value => observer.notify(resource, property, value),
+        () => {
+          throw "err";
+        }
+      );
+    }
+  };
+  return host;
+}
+```
+
+[JW-NOTE: The Resource/Host model is definitely on open design question so any guidance or alternate design feedback would be great here. In particular, I'm unsure of the type for `Host`. In many ways I like the simplicity of `{ notify: (resource: Resource, property: string, value: CalcValue) => void; }` but I'm unsure about having to pass the value down. How useful is that value, given that we probably need to go off and request all the other dependencies to do a recalc anyway. We can omit the value and encode the observer-stye signature, but at the cost of making an additional request to the resource.]
+
+##### Dynamic Dependencies
+
+Suppose we have the following formula in a text segment. `=IF(A1 < 10, Table1.ColA.Sum, Map2.Pin1.X)` where `A1 = 5`. On the first calc we request `Table1.ColA.Sum`. Suppose we change `A1 = 15` and therefore invalidate the formula. On the second calc we request `Map2.Pin1.X` but do not need `Table1.ColA.Sum`. What happens to the legacy dependency edge that links `Table1.ColA.Sum` to the formula in the text?
+
+The table has no knowledge that the formula was recalculated, so if the table changes it will send a notification to the formula. What do we do in this situation? Possible solutions:
+
+- Just recalculate the formula again. This is unnecessary work and incorrect in the presence of side-effects. This still leaves the issue of the legacy dependency. We could keep it around, and occasionally garbage collect by deleting dependencies after a notification. We could also delete after every notification, but this introduces high churn on dependency tracking.
```

Comment:
Yes, I agree that formulas should be (proper) functions. As it stands, they are functional to the extent that the initial producer (or context) that they receive is functional. As an example, take time as a producer which has the property "NOW". If we have 10 formulas hosted in a single conumer that depend on Time.NOW, then depending on when we evaluate them (and correspondingly request the NOW property), we may get different results back.

There is no current facility to freeze the context (root producer) at some point in time so that we can evaluate a batch of formulas that depend on the same producer, ensuring that we get the same results. Is this what you meant?

So essentially we want a react style ability to have some state that is immutable from the point of view of a set of consumers, and it doesn't matter when the consumers actually use the state because it won't change under their feet.

I think this where the push vs pull model comes into play. Ideally we could just push down the state to all consumers, knowing it wont change. I think this is hard with dynamic dependencies (like IF statements), and even harder when the entire programs themselves are dynamic (user entered). I think it's doable though: are there any known solutions to this problem?

One thing that we should strive for is to solve this at the consumer end rather than the producer. A simple counter producer shouldn't need to change its behaviour because its used by a huge table. One approach might be to wrap the context producer in another producer (and consumer) that effectively buffers any updates. So we would wrap the time producer in another producer/consumer that caches the first time value, and ignores and updates from the real time value. This way all 10 formulas that request Time.NOW only see the first value. Effectively this is a memoizing consumer/producer that only ever requests a property once in a given 'transaction'.